Intro to quam (#353)
qua-platform/qua-libs: Tutorials/intro-to-quam/trapped_ion/readme.md

        # We declare a QUA variable to store the boolean result of thresholding the I value.
        qubit_state = declare(int)
        # Since |1> is shelved, high fluorescence corresponds to |0>
        # i.e. I < self.threshold implies |1> and vice versa
        assign(qubit_state, Cast.to_int(I < self.threshold))
        return qubit_state

@quam_dataclass
class SingleXMacro(QubitMacro):
    def apply(self, qubit_idx: int):
        self.qubit.ion_displacement.play("ttl")
        align()
        with switch_(qubit_idx):
            with case_(1):
                self.qubit.global_mw.play("x180")
            with case_(2):
                self.qubit.global_mw.play("y180")
                self.qubit.global_mw.play("x180")
                self.qubit.global_mw.play("y180")
        align()
        self.qubit.ion_displacement.play("ttl")
        align()
```
<details>
<summary>⚠️ <code>measure_integrated</code> is a custom function added to <code>InOutSingleChannel</code> following the code below. The default <code>measure</code> implemented in QUAM performs demodulation at the IF frequency which is unnecessary for fluorescence measurement.
</summary>

> Instead, we implemented the integration of the TTL photon counting singal of the PMT to obtain the total fluorescence count. When the fluorescence exceeds a threshold in a pre-calibrated timeframe, we can discriminate the state.

Reviewer: ```suggestion
Instead, we implemented the integration of the TTL photon counting signal of the PMT to obtain the total fluorescence count. When the fluorescence exceeds a threshold in a pre-calibrated time frame, we can discriminate the state.
```

requires-python = ">=3.9"

[build-system]
requires = ["hatchling"]
build-backend = "hatchling.build"

[tool.hatch.build.targets.wheel]
packages = ["trapped_ion"]
```
</details>

While it is possible to skip this step, the saved state JSON can not be loaded for the custom components.

## 4. Pulses and channels initialization
With the data structure of our setup defined, we can now initialize the QUAM object, starting from the container.

```python
machine = Quam()
```

Another feature of QUAM is the scalability, here we will show that for $N$ qubits, we can simply run a python loop for generating a readable configuration that associates the quantum component to its corresponding hardware and pulse.

```python
n_qubits = 2
aom_position = np.linspace(200e6, 300e6, n_qubits)

# for each qubit
for i in range(n_qubits):
    qubit_id = f"q{i + 1}"
    qubit = HyperfineQubit(
        id=f"{qubit_id}",
        readout=InOutSingleChannel(
            opx_output=LFFEMAnalogOutputPort("con1", 1, 2),
            opx_input=LFFEMAnalogInputPort("con1", 1, 2),
            intermediate_frequency=aom_position[i],
        ),
        shelving=SingleChannel(

>             opx_output=LFFEMAnalogOutputPort("con1", 1, 3),

Author: Yes. For the case of our setup. I will add a mention in the text.

The $(X_\pi)^{2N}$ gate is implemented in the following with `amp_scale` is the amplitude multiplication factor on top of the original set value, `XX_rep` defined as $N$. These two variables can be parsed through the `apply` method for the `QubitMacro`.

```python
@quam_dataclass
class DoubleXMacro(QubitMacro):
    def apply(self, qubit_idx: int, amp_scale=1, XX_rep=1):
        i = declare(int)
        self.qubit.ion_displacement.play("ttl")
        align()
        with switch_(qubit_idx):
            with case_(1):
                with for_(i, 0, i < XX_rep * 2, i + 1):
                    self.qubit.global_mw.play("x180", amplitude_scale=amp_scale)
            with case_(2):
                with for_(i, 0, i < XX_rep * 2, i + 1):
                    self.qubit.global_mw.play("y180", amplitude_scale=amp_scale)
                    self.qubit.global_mw.play("x180", amplitude_scale=amp_scale)
                    self.qubit.global_mw.play("y180", amplitude_scale=amp_scale)
        align()
        self.qubit.ion_displacement.play("ttl")
        align()

machine.global_op.macros["N_XX"] = DoubleXMacro()
```

> The implemented program is similar to the case of a single addressing gate with the addition of amplitude scan range.

Reviewer: ```suggestion
The implemented program is similar to the case of a single addressing gate with the addition of amplitude scan.
```

                state = qubit.apply("measure")
                save(state, state_st)
                align()
                wait(1_000)

    with stream_processing():
        state_st.buffer(n_qubits).buffer(len(amp_scan)).average().save_all("state")
```

In the `stream_processing`, we obtain an 2D matrix corresponding to the final state for different amplitude factor and the qubits. Choosing the best amplitude that lead to $I$ in the observation, we can update the QUAM directly without the need of recording the data manually.

```python
state = job.result_handles.get("state").fetch_all()["value"]
state_scan = state[0, :, optimize_qubit_idx - 1]
best_amp_scan = amp_scan[np.argmin(state_scan)]
original_amplitude = machine.global_op.global_mw.operations["x180"].amplitude
machine.global_op.global_mw.operations["x180"].amplitude *= best_amp_scan
machine.global_op.global_mw.operations["y180"].amplitude *= best_amp_scan
```

The optimal value can be saved and loaded in the future experiment which makes organization of the experiment parameters intuitive.

```python
machine.save("state_after.json")

# Load the QUAM configuration
machine = Quam.load("state_after.json")
```

# Last word

> In this tutorial, we realized the abstraction of a physical experiment using QUAM for an operation-centric and readable structure. For physicist, this is an ideal framework for describing an experiment and book-keeping. The implementation of the code here is found in the [`machine.py`](machine.py) which serves as an entry point.

Reviewer: ```suggestion
In this tutorial, we realized the abstraction of a physical experiment using QUAM for an operation-centric and readable structure. For physicists, this is an ideal framework for describing an experiment and book-keeping. The implementation of the code here is found in the [`machine.py`](machine.py) which serves as an entry point.
```
